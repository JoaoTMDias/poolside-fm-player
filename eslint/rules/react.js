/**
 * React specific rules
 *
 * @author Jo√£o Dias (joao.dias@wit-software.com)
 * @since 1.0.0
 */
module.exports = {
	extends: ['react-app', 'airbnb'],
	plugins: ['react', 'react-hooks'],
	rules: {
		'react/jsx-wrap-multilines': 0,
		'react/jsx-closing-bracket-location': 0,
		'react/jsx-curly-spacing': 0,
		'react/jsx-equals-spacing': 0,
		'react/jsx-first-prop-new-line': 0,
		'react/jsx-indent': 0,
		'react/jsx-indent-props': 0,
		'react/jsx-max-props-per-line': 0,
		'react/jsx-tag-spacing': 0,
		'react/jsx-closing-tag-location': 0,
		'react/jsx-filename-extension': [
			'error',
			{
				extensions: ['.js', '.jsx', '.ts', '.tsx'],
			},
		],
		'react-hooks/rules-of-hooks': 'error',
		'react-hooks/exhaustive-deps': 'warn',
		'react/prop-types': 'off',
		'react/require-default-props': 'off',

		'jsx-a11y/href-no-hash': 'off',
		'jsx-a11y/anchor-is-valid': [
			'warn',
			{
				aspects: ['invalidHref'],
			},
		],
	},
};
