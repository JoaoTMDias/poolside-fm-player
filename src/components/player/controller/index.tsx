import React, { useRef, useState, useReducer } from "react";
import SoundCloudAudio from "soundcloud-audio";
import { PoolsidePlaylists } from "data/constants";
import {
	PlayerControllerContext,
	defaultPlayerControllerState,
	IPlayerControllerContext,
} from "contexts/player-controller-context";
import { useDidMount } from "helpers/custom-hooks/use-lifecycle-hooks";
import {
	ISoundcloudPlayer,
	EPlayingStatus,
	ISoundcloudPlaylist,
	ESoundCloudPlayerEvents,
	IMediaPlayerTrackMetadata,
} from "../media-player/player.interfaces";
import { updateTrackReducer } from "./update-tracker-reducer";
import { updateStatusReducer } from "./update-status-reducer";
import { isNil } from "helpers";

/**
 * Get a random number between a range
 *
 * @param {number} allTracks
 * @returns
 */
export function getRandomTrackIndex(allTracks: number) {
	const min = Math.ceil(0);
	const max = Math.floor(allTracks);
	return Math.floor(Math.random() * (max - min)) + min;
}

export interface IPlayerControllerProps {
	onPrevious?: (hasChanged: boolean) => void;
	onNext?: (hasChanged: boolean) => void;
	onPlay?: (status: EPlayingStatus) => void;
	onChangeVolume?: (volume: number) => void;
	children: React.ReactNode;
}

/**
 * PlayerController
 *
 * Controls the player instance.
 * Plays, pauses, changes tracks and then updates the state and sends it through the Context API.
 *
 */
const PlayerController: React.FunctionComponent<IPlayerControllerProps> = ({
	onPrevious,
	onNext,
	onPlay,
	onChangeVolume,
	children,
}) => {
	const { current: player } = useRef<ISoundcloudPlayer>(
		new SoundCloudAudio(process.env.REACT_APP_API_KEY)
	);
	const [artist, setArtist] = useState(defaultPlayerControllerState.artist);
	const [title, setTitle] = useState(defaultPlayerControllerState.title);
	const [duration] = useState(defaultPlayerControllerState.duration);
	const [currentTime] = useState(defaultPlayerControllerState.currentTime);
	const [currentIndex, setCurrentIndex] = useState(defaultPlayerControllerState.currentIndex);
	const [track, updateTrack] = useReducer(updateTrackReducer, defaultPlayerControllerState.track);
	const [status, updateStatus] = useReducer(
		updateStatusReducer,
		defaultPlayerControllerState.status
	);

	/**
	 * Handles the click on the previous button
	 *
	 * @returns {boolean}
	 */
	function _onPrevious(): boolean {
		let hasChanged = false;

		if (player && track) {
			const { current, last } = track;
			const previousIndex = current <= 0 ? last : current - 1;

			updateTrack({
				type: "CURRENT",
				payload: previousIndex,
			});

			if (player?.playing && player?.previous) {
				player.previous();
			}

			hasChanged = true;
		}

		if (onPrevious) {
			onPrevious(hasChanged);
		}

		return hasChanged;
	}

	/**
	 * Handles the click on the next button
	 *
	 * @returns {boolean}
	 */
	function _onNext(): boolean {
		let hasChanged = false;

		if (player && track) {
			const { current, last } = track;

			const nextIndex = current === last ? 0 : current + 1;

			updateTrack({
				type: "CURRENT",
				payload: nextIndex,
			});

			if (player?.next && player?.playing) {
				player.next();
			}

			hasChanged = true;
		}

		if (onNext) {
			onNext(hasChanged);
		}

		return hasChanged;
	}

	/**
	 * Handles the new status.
	 *
	 * @returns {void}
	 */
	function handleNewStatus(): void {
		if (player && player.play && player.pause) {
			switch (status) {
				case EPlayingStatus.playing:
					player.pause();
					break;

				case EPlayingStatus.ready:
				case EPlayingStatus.paused:
					player.play({
						playlistIndex: track.current,
					});
					break;

				default:
					break;
			}
		}
	}

	/**
	 * Callback for handling the clicking on the play button
	 *
	 * @returns {EPlayingStatus | null}
	 */
	function _onPlay(): void {
		switch (status) {
			case EPlayingStatus.ready:
			case EPlayingStatus.paused:
				updateStatus({
					type: "PLAY",
				});
				break;

			case EPlayingStatus.playing:
				updateStatus({
					type: "PAUSE",
				});
				break;

			default:
			case EPlayingStatus.idle:
			case EPlayingStatus.loading:
			case EPlayingStatus.error:
				break;
		}

		handleNewStatus();

		if (onPlay) {
			onPlay(status);
		}
	}

	/**
	 * Changes the volume of the current playing audio
	 *
	 * @param {number} volume
	 */
	function _onChangeVolume(volume: number) {
		const { audio } = player;
		if (player && audio && volume) {
			audio.volume = volume;

			if (onChangeVolume) {
				onChangeVolume(volume);
			}
		}
	}

	/**
	 * Returns the current tracks title and artist
	 *
	 * @param {number} index
	 * @returns {(IMediaPlayerTrackMetadata | null)}
	 */
	function getCurrentTrackAndArtist(index: number): IMediaPlayerTrackMetadata | null {
		if (index && player?._playlist) {
			// eslint-disable no-underscore-dangle
			const { title: currentTitle } = player._playlist.tracks[index];
			const { user } = player._playlist.tracks[index];
			// eslint-enable no-underscore-dangle

			const currentArtist = user && user.username;

			if (title && artist) {
				return {
					title: currentTitle,
					artist: currentArtist,
				};
			}
		}

		return null;
	}

	/**
	 * Loads a new playlist
	 *
	 * @param {ISoundcloudPlaylist} playlist
	 */
	async function loadPlaylist(playlist: ISoundcloudPlaylist) {
		const current = getRandomTrackIndex(playlist.track_count);
		const last = playlist.track_count - 1;

		updateStatus({
			type: "LOAD",
		});

		if (player && player.on && player) {
			// once playlist is loaded it can be played
			updateTrack({
				type: "TRACK",
				payload: {
					current,
					last,
				},
			});

			updateStatus({
				type: "READY",
			});

			player.on(ESoundCloudPlayerEvents.canplay, () => {
				const { _playlistIndex } = player;
				const metadata = _playlistIndex ? getCurrentTrackAndArtist(_playlistIndex) : null;

				if (metadata) {
					setTitle(metadata.title);
					setArtist(metadata.artist);
				}
			});

			// for playlists it's possible to switch to another track in queue
			// e.g. we do it here when playing track is finished
			player.on(ESoundCloudPlayerEvents.ended, () => {
				_onNext();
			});
		}
	}

	/**
	 * get the current playlist and starts a new playlist instance
	 *
	 * @returns {void}
	 */
	function startPlaylist(): void {
		const currentPlaylist = PoolsidePlaylists[currentIndex].url;

		if (player?.resolve && currentPlaylist) {
			player.resolve(currentPlaylist, (playlist: ISoundcloudPlaylist) => {
				loadPlaylist(playlist);
			});
		}
	}

	/**
	 * When the user selects a new playlist, updates the current playlist index state
	 *
	 * @param {number} index
	 * @returns {void}
	 */
	function _onChangeOption(index: number): void {
		if (!isNil(index) && player && player.stop) {
			player.stop();

			setCurrentIndex(index);
			startPlaylist();
		}
	}

	useDidMount(() => {
		if (player) {
			startPlaylist();
		}
	});

	const contextValue: IPlayerControllerContext = {
		status,
		artist,
		title,
		duration,
		currentTime,
		currentIndex,
		track,
		audio: player.audio,
		previous: () => _onPrevious(),
		next: () => _onNext(),
		togglePlay: () => _onPlay(),
		changeVolume: (volume) => _onChangeVolume(volume),
		onChangeOption: (index) => _onChangeOption(index),
	};

	return (
		<PlayerControllerContext.Provider value={contextValue}>
			{children}
		</PlayerControllerContext.Provider>
	);
};

export default PlayerController;
