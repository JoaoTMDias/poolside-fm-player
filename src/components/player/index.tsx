import ControlsMedia from "./controls-media/index";
import AudioPlayer from "./audio-player/index";

export { AudioPlayer, ControlsMedia };
