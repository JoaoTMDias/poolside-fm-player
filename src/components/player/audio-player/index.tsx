/* eslint-disable jsx-a11y/media-has-caption */
// Libraries
import * as React from "react";
import ControlsMedia from "../controls-media/index";
import { IAudioPlayerProps, IAudioPlayerState, EPlayingStatus } from "../player.interfaces";

/**
 * @description Component Description
 * @author Jo√£o Dias
 * @date 2019-02-16
 * @returns {React.FunctionComponent<IAudioPlayerProps>}
 */
class AudioPlayer extends React.Component<IAudioPlayerProps, IAudioPlayerState> {
	private player: React.RefObject<HTMLAudioElement>;

	constructor(props: IAudioPlayerProps) {
		super(props);

		this.state = {
			player: null,
			status: EPlayingStatus.paused,
		};

		// Binds
		this.onClickOnPrevious = this.onClickOnPrevious.bind(this);
		this.onClickOnPlay = this.onClickOnPlay.bind(this);
		this.onClickOnNext = this.onClickOnNext.bind(this);

		// Refs
		this.player = React.createRef<HTMLAudioElement>();
	}

	/**
	 *
	 *
	 * @memberof AudioPlayer
	 */
	componentDidMount() {
		const { player } = this.state;
		if (this.player.current) {
			this.setState(
				{
					player: this.player.current,
				},
				() => {
					if (player) {
						player.currentTime = 0;
						player.pause();
					}
				},
			);
		}
	}

	/**
	 *
	 *
	 * @memberof AudioPlayer
	 */
	onClickOnPrevious() {
		console.log("previous");
	}

	/**
	 *
	 *
	 * @memberof AudioPlayer
	 */
	onClickOnNext() {
		console.log("next");
	}

	/**
	 *
	 *
	 * @memberof AudioPlayer
	 */
	onClickOnPlay() {
		const { status } = this.state;

		let nextStatus;

		switch (status) {
			default:
			case EPlayingStatus.paused:
				nextStatus = EPlayingStatus.playing;
				break;

			case EPlayingStatus.playing:
				nextStatus = EPlayingStatus.paused;
		}

		this.handleAudioPlayer(nextStatus);
	}

	/**
	 *
	 *
	 * @param {EPlayingStatus} status
	 * @memberof AudioPlayer
	 */
	handleAudioPlayer(status: EPlayingStatus) {
		const { player } = this.state;

		if (player) {
			switch (status) {
				case EPlayingStatus.playing:
					player.play();
					break;

				case EPlayingStatus.paused:
					player.pause();
					break;

				default:
					break;
			}

			this.updatePlayingStatus(status);
		}
	}

	/**
	 *
	 *
	 * @param {EPlayingStatus} status
	 * @memberof AudioPlayer
	 */
	updatePlayingStatus(status: EPlayingStatus) {
		this.setState({
			status,
		});
	}

	render() {
		const { status } = this.state;
		const { src } = this.props;

		return (
			<div id="media-player" className="media-player row">
				<audio
					ref={this.player}
					id="audio-player"
					src={src}
					data-testid="component-audio-player"
					autoPlay={false}
					className="sr-only"
				/>
				<ControlsMedia
					status={status}
					onClickOnPrevious={this.onClickOnPrevious}
					onClickOnPlay={this.onClickOnPlay}
					onClickOnNext={this.onClickOnNext}
				/>
			</div>
		);
	}
}

export default React.memo(AudioPlayer);
