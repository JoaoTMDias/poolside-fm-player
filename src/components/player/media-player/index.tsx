/* eslint-disable jsx-a11y/media-has-caption */
// Libraries
import * as React from "react";
import { PlayerControllerContext } from "contexts/player-controller-context";
import CurrentSong from "../../current-song/index.component";
import ControlsMedia from "../controls-media/index";

/**
 * @description Component Description
 * @author Jo√£o Dias
 * @date 2019-02-16
 * @returns {React.FunctionComponent<IMediaPlayerProps>}
 */
const MediaPlayer = () => {
	return (
		<PlayerControllerContext.Consumer>
			{({
				artist,
				changeVolume,
				currentTime,
				next,
				previous,
				status,
				title,
				togglePlay,
				track,
			}) => {
				return (
					<div
						id="media-player"
						data-testid="component-media-player"
						className="media-player row"
						data-current-track={track?.current.toString()}
						data-playing-status={status}
					>
						<CurrentSong title={title} artist={artist} currentTime={`${currentTime}`} />
						<ControlsMedia
							status={status}
							onClickOnPrevious={previous}
							onTogglePlay={togglePlay}
							onClickOnNext={next}
							onChangeVolume={changeVolume}
						/>
					</div>
				);
			}}
		</PlayerControllerContext.Consumer>
	);
};

export default MediaPlayer;
