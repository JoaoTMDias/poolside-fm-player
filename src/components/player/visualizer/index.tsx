// Libraries
import * as React from "react";
import { PlayerControllerContext } from "contexts/player-controller-context";
import { EPlayingStatus } from "components/player/media-player/player.interfaces";
import produce from "immer";
import * as S from "./player-visualizer.styled";

export interface IPlayerVisualizerProps {
	audio: HTMLAudioElement | null;
	status: EPlayingStatus;
	devicePixelRatio?: number | null;
}
export interface IPlayerVisualizerState {
	width: number;
	height: number;
	bars: {
		total: number;
		background: string;
		color: string;
	};
}

export const CAPTIONS = {
	idle: "Press play to start",
	paused: "Press play to resume",
	error: "An error occurred",
	loading: "Loading tape...",
	ready: "Ready!",
	playing: "Music is Playing"
}

/**
 * @description An audio visualizer for the player
 * @author Jo√£o Dias
 * @date 2019-02-16
 * @returns {React.FunctionComponent}
 */
class PlayerVisualizer extends React.Component<IPlayerVisualizerProps, IPlayerVisualizerState> {
	static contextType = PlayerControllerContext;

	private canvas: React.RefObject<HTMLCanvasElement>;

	private canvasElement: HTMLCanvasElement | null;

	public canvasContext: CanvasRenderingContext2D | null;

	private audioContext: AudioContext | null;

	private audioAnalyser: AnalyserNode | null;

	private mediaElementSource: MediaElementAudioSourceNode | null;

	private playerBarWidth: number;

	private dataArray: Uint8Array | null;

	private bufferLength: number;

	context!: React.ContextType<typeof PlayerControllerContext>;

	static defaultProps: IPlayerVisualizerProps = {
		audio: null,
		status: EPlayingStatus.idle,
		devicePixelRatio: window.devicePixelRatio
	};

	constructor (props: IPlayerVisualizerProps) {
		super(props);

		this.canvas = React.createRef<HTMLCanvasElement>();
		this.canvasElement = null;
		this.canvasContext = null;
		this.audioContext = null;
		this.audioAnalyser = null;
		this.mediaElementSource = null;
		this.playerBarWidth = 0;
		this.dataArray = null;
		this.bufferLength = 0;

		this.state = {
			width: 256,
			height: 32,
			bars: {
				total: 128,
				background: "#000000",
				color: "white",
			},
		};
	}

	/**
	 * - Gets the black and white css colors from the html element
	 * - Updates the state with the canvas element reference
	 * - Initializes the canvas drawing
	 *
	 * @memberof PlayerVisualizer
	 */
	componentDidMount() {
		const { audio, devicePixelRatio } = this.props;
		const { width, height } = this.state;

		const dpi = devicePixelRatio || 1;

		this.getColorsFromRoot();

		if (this.canvas && this.canvas.current) {
			this.canvasElement = this.canvas.current;

			this.setCanvasPixelRatio(width, dpi, height);

			if (audio && audio.src.length > 0) {
				this.initCanvas(audio);
			}
		}
	}

	/**
	 * Only updathes the component when props change
	 *
	 * @param {IPlayerVisualizerProps} nextProps
	 * @returns
	 * @memberof PlayerVisualizer
	 */
	shouldComponentUpdate(nextProps: IPlayerVisualizerProps) {
		const { audio, status } = this.props;
		if (nextProps.audio !== audio || nextProps.status !== status) {
			return true;
		}

		return false;
	}

	/**
	 * If there's a new audio, initializes the canvas
	 *
	 * @param {IPlayerVisualizerProps} prevProps
	 * @memberof PlayerVisualizer
	 */
	componentDidUpdate() {
		const { audio } = this.props;

		if (audio && audio.src) {
			this.initCanvas(audio);

			return true;
		}

		return false;
	}

	/**
	 * Defines the canvas width and height base on the devices pixel ratio
	 *
	 * @private
	 * @param {number} width
	 * @param {number} dpi
	 * @param {number} height
	 * @memberof PlayerVisualizer
	 */
	private setCanvasPixelRatio(width: number, dpi: number, height: number) {
		if (this.canvasElement) {
			this.canvasElement.width = width * dpi;
			this.canvasElement.height = height * dpi;
			this.canvasElement.setAttribute("width", `${width}`);
			this.canvasElement.setAttribute("height", `${height}`);
		}
	}

	/**
	 * Gets the black and white css colors from the html element
	 *
	 * @memberof PlayerVisualizer
	 */
	getColorsFromRoot() {
		const background = getComputedStyle(document.documentElement)
			.getPropertyValue("--color-black")
			.trim();
		const color = getComputedStyle(document.documentElement)
			.getPropertyValue("--color-white")
			.trim();

		this.setState(
			produce((draftState: IPlayerVisualizerState) => {
				draftState.bars.background = background;
				draftState.bars.color = color.length > 0 ? color : "#faf9f9";
			})
		);
	}

	/**
	 *
	 *
	 * @returns
	 * @memberof PlayerVisualizer
	 */
	getStatusMetadata() {
		const { status } = this.props;

		const defaultState = {
			classname: "",
			caption: {
				title: "Press play to start",
				classname: "sr-only"
			}
		};

		const metadata = produce(defaultState, (draftState, action = status) => {
			// eslint-disable-next-line default-case
			switch (action) {
				case EPlayingStatus.idle:
					draftState.caption.title = CAPTIONS.idle;
					break;

				case EPlayingStatus.paused:
					draftState.caption.title = CAPTIONS.paused;
					break;

				case EPlayingStatus.error:
					draftState.caption.title = CAPTIONS.error;
					break;

				case EPlayingStatus.loading:
					draftState.caption.title = CAPTIONS.loading;
					break;

				case EPlayingStatus.ready:
					draftState.caption.title = CAPTIONS.ready;
					break;

				case EPlayingStatus.playing:
					draftState.classname = "is-playing";
					draftState.caption.title = CAPTIONS.playing;
					break;
			}

			switch (action) {
				case EPlayingStatus.error:
				case EPlayingStatus.idle:
				case EPlayingStatus.loading:
				case EPlayingStatus.ready:
				case EPlayingStatus.paused:
					draftState.caption.classname = "player-visualizer__label";
					break;

				default:
					break;
			}
		});

		return metadata;
	}

	/**
	 * Creates an object that provides methods and properties
	 * for drawing graphics on the canvas element
	 *
	 * @param {HTMLCanvasElement} canvasElement
	 * @memberof PlayerVisualizer
	 */
	createCanvasContext(canvasElement: HTMLCanvasElement) {
		const { width, height, bars } = this.state;

		this.canvasContext = canvasElement.getContext("2d");

		if (this.canvasContext) {
			this.canvasContext.fillStyle = bars.background;
			this.canvasContext.fill();
			this.canvasContext.fillRect(0, 0, width, height);
		}
	}

	/**
	 * Creates the Audio Analyser that feeds on the current audio
	 *
	 * @param {HTMLAudioElement} audio
	 * @memberof PlayerVisualizer
	 */
	createAudioAnalyser(audio: HTMLAudioElement) {
		if (!this.audioContext) {
			this.audioContext = new AudioContext();
			this.audioAnalyser = this.audioContext.createAnalyser();
			this.audioAnalyser.connect(this.audioContext.destination);
			this.audioAnalyser.fftSize = 256;
			this.bufferLength = this.audioAnalyser.frequencyBinCount;
			this.dataArray = new Uint8Array(this.bufferLength);
		}

		if (!this.mediaElementSource) {
			this.mediaElementSource = this.audioContext.createMediaElementSource(audio);
		}

		audio.crossOrigin = "anonymous";

		if (this.mediaElementSource && this.audioAnalyser) {
			this.mediaElementSource.connect(this.audioAnalyser);
		}
	}

	/**
	 * Renders the bars frames onto the canvas
	 *
	 * @memberof PlayerVisualizer
	 */
	renderFrame = () => {
		requestAnimationFrame(this.renderFrame);

		const { bars } = this.state;
		let xAxis = 0;
		let barHeight;

		if (this.canvasContext && this.dataArray && this.playerBarWidth && this.canvasElement) {
			if (this.audioAnalyser && this.audioAnalyser !== undefined) {
				this.audioAnalyser.getByteFrequencyData(this.dataArray);
			}

			this.canvasContext.fillStyle = bars.background;
			this.canvasContext.fillRect(0, 0, this.canvasElement.width, this.canvasElement.height);

			for (let i = 0; i < this.bufferLength; i += 1) {
				barHeight = this.dataArray[i] * 0.1;

				this.canvasContext.fillStyle = bars.color;
				this.canvasContext.fillRect(
					xAxis,
					this.canvasElement.height - barHeight,
					this.playerBarWidth,
					barHeight
				);

				xAxis += this.playerBarWidth + 2;
			}
		}
	};

	/**
	 * Initializes the Canvas creation.
	 * 1- Creates the player context
	 * 2- Creates an analyser based on the current audio
	 * 3- Renders the frames onto the canvas
	 *
	 * @param {HTMLCanvasElement} player
	 * @memberof PlayerVisualizer
	 */
	initCanvas(audio: HTMLAudioElement) {
		if (this.canvasElement) {
			this.createCanvasContext(this.canvasElement);

			this.createAudioAnalyser(audio);

			this.playerBarWidth = 1;

			this.renderFrame();
		}
	}

	render() {
		const { audio, devicePixelRatio, status } = this.props;
		const metadata = this.getStatusMetadata();

		return (
			<S.PlayerVisualizerWrapper
				role="presentation"
				data-testid="player-visualizer-wrapper"
				id="player-visualizer"
				className={`player-visualizer ${metadata.classname}`}
			>
				<figcaption id="player-visualizer-label" data-testid="component-player-visualizer-label" className={metadata.caption.classname}>
					{metadata.caption.title}
				</figcaption>
				<div
					className="sr-only"
					data-testid="player-visualizer-meta"
					data-status={status}
					data-audio={audio?.src}
					data-devicepixelratio={`${devicePixelRatio}`}
				/>
				<canvas
					ref={this.canvas}
					id="player-visualizer-canvas"
					data-testid="component-player-visualizer-canvas"
					aria-labelledby="player-visualizer-label"
					width="272"
					height="40"
					className="player-visualizer__canvas"
				/>
			</S.PlayerVisualizerWrapper>
		);
	}
}

export default PlayerVisualizer;
