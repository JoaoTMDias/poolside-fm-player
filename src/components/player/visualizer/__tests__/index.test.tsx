import React from "react";
import Audio from "mock-audio-element";
import { render, cleanup } from "@testing-library/react";
import { mount, ReactWrapper } from "enzyme";
import PlayerVisualizer, {
	CAPTIONS,
	IPlayerVisualizerProps,
	IPlayerVisualizerState,
} from "components/player/visualizer/index";
import { EPlayingStatus } from "components/player/media-player/player.interfaces";
import { createCanvas } from "canvas";

afterEach(cleanup);

const getMockedAudioElement = (src?: string) => {
	const mock: HTMLAudioElement = new Audio();

	mock.crossOrigin = "anonymous";

	mock.src = src || "https://www.soundhelix.com/examples/mp3/SoundHelix-Song-1.mp3";

	return mock;
};

type VisualizerType = ReactWrapper<IPlayerVisualizerProps, IPlayerVisualizerState>;

describe("<PlayerVisualizer />", () => {
	describe("render", () => {
		it("should render a visualizer", () => {
			const component = render(<PlayerVisualizer status={EPlayingStatus.paused} />);

			expect(component).toMatchSnapshot();
		});
	});

	describe("captions", () => {
		function setup(status: EPlayingStatus) {
			const { getByTestId } = render(<PlayerVisualizer status={status} />);

			return {
				visualizer: getByTestId("player-visualizer-wrapper"),
				label: getByTestId("component-player-visualizer-label"),
			};
		}

		it("should render the metadata when the status is idle", () => {
			const { visualizer, label } = setup(EPlayingStatus.idle);

			expect(label.textContent).toBe(CAPTIONS.idle);
			expect(label.classList.contains("player-visualizer__label")).toBe(true);
			expect(label.classList.contains("sr-only")).toBe(false);
			expect(visualizer.classList.contains("is-playing")).toBe(false);
		});

		it("should render the metadata when the status is paused", () => {
			const { visualizer, label } = setup(EPlayingStatus.paused);

			expect(label.textContent).toBe(CAPTIONS.paused);
			expect(label.classList.contains("player-visualizer__label")).toBe(true);
			expect(label.classList.contains("sr-only")).toBe(false);
			expect(visualizer.classList.contains("is-playing")).toBe(false);
		});

		it("should render the metadata when the status has an error", () => {
			const { visualizer, label } = setup(EPlayingStatus.error);

			expect(label.textContent).toBe(CAPTIONS.error);
			expect(label.classList.contains("player-visualizer__label")).toBe(true);
			expect(label.classList.contains("sr-only")).toBe(false);
			expect(visualizer.classList.contains("is-playing")).toBe(false);
		});

		it("should render the metadata when the status is loading", () => {
			const { visualizer, label } = setup(EPlayingStatus.loading);

			expect(label.textContent).toBe(CAPTIONS.loading);
			expect(label.classList.contains("player-visualizer__label")).toBe(true);
			expect(label.classList.contains("sr-only")).toBe(false);
			expect(visualizer.classList.contains("is-playing")).toBe(false);
		});

		it("should render the metadata when the status is ready", () => {
			const { visualizer, label } = setup(EPlayingStatus.ready);

			expect(label.textContent).toBe(CAPTIONS.ready);
			expect(label.classList.contains("player-visualizer__label")).toBe(true);
			expect(label.classList.contains("sr-only")).toBe(false);
			expect(visualizer.classList.contains("is-playing")).toBe(false);
		});

		it("should render the metadata when the status is playing", () => {
			const { visualizer, label } = setup(EPlayingStatus.playing);

			expect(label.textContent).toBe(CAPTIONS.playing);
			expect(label.classList.contains("player-visualizer__label")).toBe(false);
			expect(label.classList.contains("sr-only")).toBe(true);
			expect(visualizer.classList.contains("is-playing")).toBe(true);
		});
	});

	describe("updates", () => {
		it("should update when status changes", () => {
			const { getByTestId, rerender } = render(<PlayerVisualizer status={EPlayingStatus.paused} />);

			const status = getByTestId("player-visualizer-meta").getAttribute("data-status");

			expect(status).toBe("paused");

			rerender(<PlayerVisualizer status={EPlayingStatus.loading} />);

			const newStatus = getByTestId("player-visualizer-meta").getAttribute("data-status");

			expect(newStatus).toBe("loading");
			expect(newStatus !== status).toBe(true);
		});
	});
});
