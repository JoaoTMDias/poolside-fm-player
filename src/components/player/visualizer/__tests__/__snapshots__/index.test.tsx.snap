// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlayerVisualizer /> render should render a visualizer 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <figure
        class="sc-AxjAm jpoiFA player-visualizer "
        data-testid="player-visualizer-wrapper"
        id="player-visualizer"
        role="presentation"
      >
        <figcaption
          class="player-visualizer__label"
          data-testid="component-player-visualizer-label"
          id="player-visualizer-label"
        >
          Press play to resume
        </figcaption>
        <div
          class="sr-only"
          data-devicepixelratio="1"
          data-status="paused"
          data-testid="player-visualizer-meta"
        />
        <canvas
          aria-labelledby="player-visualizer-label"
          class="player-visualizer__canvas"
          data-testid="component-player-visualizer-canvas"
          height="32"
          id="player-visualizer-canvas"
          width="256"
        />
      </figure>
    </div>
  </body>,
  "container": <div>
    <figure
      class="sc-AxjAm jpoiFA player-visualizer "
      data-testid="player-visualizer-wrapper"
      id="player-visualizer"
      role="presentation"
    >
      <figcaption
        class="player-visualizer__label"
        data-testid="component-player-visualizer-label"
        id="player-visualizer-label"
      >
        Press play to resume
      </figcaption>
      <div
        class="sr-only"
        data-devicepixelratio="1"
        data-status="paused"
        data-testid="player-visualizer-meta"
      />
      <canvas
        aria-labelledby="player-visualizer-label"
        class="player-visualizer__canvas"
        data-testid="component-player-visualizer-canvas"
        height="32"
        id="player-visualizer-canvas"
        width="256"
      />
    </figure>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
